File: backend/pkg/security/policies/threshold.py
from __future__ import annotations

from pkg.apis.security.models import AggregatedDetectionResult, SecurityDecision


class ThresholdPolicy:
    """score >= threshold → BLOCK, otherwise ALLOW.

    Threshold is injected from configuration so detector implementations stay
    threshold-agnostic. This is required for TPR @ target-FPR sweeps.
    """

    name = "threshold"

    def __init__(self, threshold: float, *, action_on_detection: str = "block") -> None:
        if not 0.0 <= threshold <= 1.0:
            raise ValueError("threshold must be between 0.0 and 1.0")
        self.threshold = threshold
        self.action_on_detection = action_on_detection.lower()

    def evaluate(self, result: AggregatedDetectionResult) -> SecurityDecision:
        if result.score >= self.threshold:
            if self.action_on_detection == "review":
                return SecurityDecision.REVIEW
            return SecurityDecision.BLOCK
        return SecurityDecision.ALLOW

File: backend/pkg/security/policies/threshold.go
package policies

import "github.com/spider/spider/pkg/apis"

type Policy interface {
	Name() string
	Threshold() float64
	Evaluate(result apis.AggregatedDetectionResult) apis.SecurityDecision
}

type ThresholdPolicy struct {
	ThresholdValue     float64
	ActionOnDetection  string
}

func NewThresholdPolicy(threshold float64, actionOnDetection string) *ThresholdPolicy {
	if actionOnDetection == "" {
		actionOnDetection = "block"
	}
	return &ThresholdPolicy{ThresholdValue: threshold, ActionOnDetection: actionOnDetection}
}

func (p *ThresholdPolicy) Name() string { return "threshold" }

func (p *ThresholdPolicy) Threshold() float64 { return p.ThresholdValue }

func (p *ThresholdPolicy) Evaluate(result apis.AggregatedDetectionResult) apis.SecurityDecision {
	if result.Score >= p.ThresholdValue {
		if p.ActionOnDetection == "review" {
			return apis.DecisionReview
		}
		return apis.DecisionBlock
	}
	return apis.DecisionAllow
}
